app/src/main/java/org/wikipedia/places/PlacesFragment.kt

```diff
@@ -485,6 +509,100 @@ class PlacesFragment : Fragment(), MapboxMap.OnMapClickListener {
+                        moveToReadingList(readingListPage.listId, pageTitle, true)
+                    }
+                }
+
+                override fun onRemoveRequest() {
+                    ExclusiveBottomSheetPresenter.dismiss(childFragmentManager)
+                }
+            }).show(HistoryEntry(pageTitle, HistoryEntry.SOURCE_PLACES))
+        } else {
+            addToReadingList(pageTitle, true)
+        }
+    }
+
+    private fun addToReadingList(title: PageTitle, addToDefault: Boolean) {
+        if (addToDefault) {
+            // If the title is a redirect, resolve it before saving to the reading list.
+            lifecycleScope.launch(CoroutineExceptionHandler { _, t -> L.e(t) }) {
+                var finalPageTitle = title
+                try {
+                    ServiceFactory.get(title.wikiSite).getInfoByPageIdsOrTitles(null, title.prefixedText)
+                        .query?.firstPage()?.let {
+                            finalPageTitle = PageTitle(it.title, title.wikiSite, it.thumbUrl(), it.description, it.displayTitle(title.wikiSite.languageCode), null)
+                        }
+                } finally {
+                    ReadingListBehaviorsUtil.addToDefaultList(requireActivity(), finalPageTitle, Constants.InvokeSource.PLACES) { readingListId ->
+                        moveToReadingList(readingListId, finalPageTitle, false) }
+                }
+            }
+        } else {
+            ExclusiveBottomSheetPresenter.showAddToListDialog(childFragmentManager, title, Constants.InvokeSource.PLACES) { }
+        }
+    }
+
+    private fun moveToReadingList(sourceReadingListId: Long, title: PageTitle, showDefaultList: Boolean) {
+        ExclusiveBottomSheetPresenter.showMoveToListDialog(childFragmentManager, sourceReadingListId, title, Constants.InvokeSource.PLACES, showDefaultList) { }
+    }
+
+    override fun onLinkPreviewLoadPage(title: PageTitle, entry: HistoryEntry, inNewTab: Boolean) {
+        if (inNewTab) {
+            TabUtil.openInNewBackgroundTab(entry)
+            // TODO: run animation
+            requireActivity().invalidateOptionsMenu()
+        } else {
+            startActivity(PageActivity.newIntentForCurrentTab(requireActivity(), entry, entry.title, false))
+        }
+    }
+
+    override fun onLinkPreviewWatch(title: PageTitle, lastWatchExpiry: WatchlistExpiry, isWatched: Boolean) {
+        viewModel.watchlistExpiryChanged = false
+        viewModel.watchOrUnwatch(lastWatchExpiry, isWatched)
+        ExclusiveBottomSheetPresenter.dismiss(childFragmentManager)
+    }
+
+    override fun onLinkPreviewAddToList(title: PageTitle, isInReadingList: Boolean, anchor: View?) {
+        anchor?.let {
+            onAddToReadingListClick(title, isInReadingList, it)
+        }
+    }
+    override fun onExpirySelect(expiry: WatchlistExpiry) {
+        viewModel.watchOrUnwatch(expiry, false)
+        ExclusiveBottomSheetPresenter.dismiss(childFragmentManager)
+    }
+
```

Comment:
Similar to previous comments, can all of this Watchlist and Reading List logic be self-contained in the Link Preview component?  (with some logic possibly going into WatchlistExpiryDialog)
This will make all of the logic reusable wherever Link Previews are shown.

Reply:
Please review the separate PR of moving the logic of changing the watchlist expiry here: #4365